### PR TITLE
HDDS-10412. Prefix ACL check needs to resolve the bucket link

### PR DESCRIPTION
**Pull request by ivandika3**
## What changes were proposed in this pull request?

Following up on [HDDS-4715](https://issues.apache.org/jira/browse/HDDS-4715), we also need to resolve the bucket link for prefix ACL.

Here are the changes included in this patch:
-  Include prefix resource type to be `BucketManagerImpl`  so that prefix access under a bucket link will be resolved to the source bucket
- Prefix ACL read: `PrefixManagerImpl#getAcl` will resolve the bucket to get the ACL
- Prefix ACL write: `OMPrefixAclRequest#validateAndUpdateCache` will resolve the bucket before writing the ACL
   - This requires passing the resolved prefix `OzoneObj` to the request hooks (`onComplete`, `apply`) so that the request implementation can use the resolved object and not the link object
- Fixed `PrefixManagerImpl#setAcl` so that default ACL is inherited only when the prefix does not exist
   - This was found in the integration test
- Some comments and cosmetics improvements

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-10412

## How was this patch tested?

Acceptance and integration tests.

Clean CI run: https://github.com/ivandika3/ozone/actions/runs/8029229022

### CODE REVIEW

**Comment by ivandika3** [comment]
@whbing @ChenSammi Could you help take a look when you have time?

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/PrefixManagerImpl.java [hunk 1]
```diff
@@ -322,23 +331,53 @@ public OMPrefixAclOpResult setAcl(OzoneObj ozoneObj, List<OzoneAcl> ozoneAcls,
...
+    if (StringUtils.isEmpty(obj.getVolumeName()) || StringUtils.isEmpty(obj.getBucketName())) {
+      return obj;
+    }
+
+    ResolvedBucket resolvedBucket = ozoneManager.resolveBucketLink(
+        Pair.of(obj.getVolumeName(), obj.getBucketName()));
+    String realVolume = resolvedBucket.realVolume();
+    String realBucket = resolvedBucket.realBucket();
+    // If the bucket is not a link bucket, just return the Ozone object
+    if (realVolume.equals(obj.getVolumeName()) && realBucket.equals(obj.getBucketName())) {
```
[the remark this answers is not in the document]
Comment (ivandika3) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/PrefixManagerImpl.java]: Thank you for the review. It's superseded by the suggested extraction in the other suggestion.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/PrefixManagerImpl.java [hunk 2]
```diff
@@ -322,23 +331,53 @@ public OMPrefixAclOpResult setAcl(OzoneObj ozoneObj, List<OzoneAcl> ozoneAcls,
...
+      return obj;
+    }
+    // If the bucket is a link bucket, return a new OzoneObj with resolved volume and bucket name
+    return OzoneObjInfo.Builder.fromOzoneObj(obj)
+        .setVolumeName(realVolume).setBucketName(realBucket).build();
+  }
```
[the remark this answers is not in the document]
Comment (ivandika3) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/PrefixManagerImpl.java]: Thank you for the suggestion. Updated. PTAL.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/prefix/OMPrefixAclRequest.java [hunk 3]
```diff
@@ -145,16 +142,21 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager, TermIn
...
+            .releaseWriteLock(PREFIX_LOCK, prefixPath));
       }
       if (omClientResponse != null) {
         omClientResponse.setOmLockDetails(getOmLockDetails());
       }
     }
 
-    OzoneObj obj = getOzoneObj();
+    OzoneObj obj = resolvedPrefixObj;
+    if (obj == null) {
```
Comment (whbing) [inline hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/prefix/OMPrefixAclRequest.java:153]: Is null condition necessary here ? I see a call to `resolvedPrefixObj.getpath ()` on line 81, 82.

Comment (ivandika3) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/prefix/OMPrefixAclRequest.java:153]: If `IOException` is thrown during the `getResolvedPrefixObj`, the `resolvedPrefixObj` should be null and we should fall back to the initial ozone object from `getOzoneObj`.

File: hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/prefix/OMPrefixAclRequest.java [hunk 4]
```diff
@@ -68,28 +66,27 @@ public OMClientResponse validateAndUpdateCache(OzoneManager ozoneManager, TermIn
     boolean lockAcquired = false;
     String volume = null;
     String bucket = null;
-    String key = null;
+    String prefix = null;
```
[the remark this answers is not in the document]
Comment (ivandika3) [reply hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/request/key/acl/prefix/OMPrefixAclRequest.java]: Thank you for the review. I removed all the unnecessary variables (`volume`, `bucket`, and `prefix`)

**Comment by whbing** [comment]
@ivandika3 Thanks ! LGTM.

**Comment by adoroszlai** [comment]
@ChenSammi would you like to take a look?

**Review by adoroszlai — APPROVED** [review]

**Comment by adoroszlai** [comment]
Thanks @ivandika3 for the patch, @whbing for the review.

**Comment by ivandika3** [comment]
Thank you @adoroszlai @whbing for the reviews.

**Pull request merged**